aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -818,6 +840,80 @@ Tensor& linalg_cholesky_out(Tensor &result, const Tensor &self) {
   return result;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ cholesky_inverse ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+DEFINE_DISPATCH(cholesky_inverse_stub);
+
+Tensor& cholesky_inverse_out_info(Tensor& result, Tensor& infos, const Tensor& input, bool upper) {
+  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT(input.size(-1) == input.size(-2));
+
+  TORCH_INTERNAL_ASSERT(result.scalar_type() == input.scalar_type());
+  TORCH_INTERNAL_ASSERT(result.device() == input.device());
+
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.device() == at::kCPU);
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
+
+  // if result has no elements we can modify it
+  if (result.numel() == 0) {
+    at::native::resize_as_(result, input.transpose(-2, -1), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);
+  }
+
+  // result tensor must be in batched column major order (Fortran contiguous)
+  TORCH_INTERNAL_ASSERT(result.transpose(-2, -1).is_contiguous());
+  TORCH_INTERNAL_ASSERT(result.sizes().equals(input.sizes()));
+
+  // cholesky_inverse_stub (apply_cholesky_inverse) performs calculations in-place and result must be a copy of input
+  result.copy_(input);
+
+  // infos must be contiguous
+  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+  infos.fill_(0);
+
+  result = cholesky_inverse_stub(result.device().type(), result, infos, upper);
+  return result;
+}
+
+Tensor& cholesky_inverse_out(const Tensor &input, bool upper, Tensor &result) {
+  squareCheckInputs(input);
+  TORCH_CHECK(result.scalar_type() == input.scalar_type(),
+    "result dtype ", result.scalar_type(), " does not match input dtype ", input.scalar_type());
+  TORCH_CHECK(result.device() == input.device(),
+    "result device ", result.device(), " does not match input device ", input.device());
+
+  // Single matrix MAGMA routine requires 'infos' to reside in CPU memory,
```

Comment:
Wait -- doesn't this PR implemented batched CUDA, though?

Reply:
Oh, this comment is a bit outdated. I need to change it. At that time I thought that batched MAGMA routine would require `infos` to live on GPU.
Now CUDA path is implemented in terms of MAGMA's `apply_cholesky_solve`. It is different from other batched functions in that it doesn't take an array of ints `infos` argument, it doesn't raise any errors related to the algorithm. It returns only a single integer to tell whether all passed arguments were ok or not.
If cuSOLVER would be used instead in the future, then `infos` would need to be created on GPU.

I will change it to:
* if input on CPU then we need `infos` of size `batchsize(input)` to fill with error codes from LAPACK for each matrix in the batched tensor.
* if input on GPU then we need only one integer living on CPU for storing error code from MAGMA

It also means that for some inputs CPU could raise an error while GPU would output something because the operations used are different.

Reply:
> It also means that for some inputs CPU could raise an error while GPU would output something because the operations used are different.

Are there some specific cases you're thinking of?

Reply:
Checked the docs. It will happen for inputs with zero on the diagonal. CPU version raises an error with the index of the zero diagonal element, while CUDA gives `inf`. I need to add this to tests.
```python
In [1]: import torch

In [2]: a = torch.randn(2, 2)

In [3]: a
Out[3]: 
tensor([[-0.6556, -0.4479],
        [-0.9347,  0.8169]])

In [4]: a[1, 1] = 0

In [5]: a
Out[5]: 
tensor([[-0.6556, -0.4479],
        [-0.9347,  0.0000]])

In [6]: torch.cholesky_inverse(a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-680561e9e67c> in <module>
----> 1 torch.cholesky_inverse(a)

RuntimeError: Lapack Error potri : A(2,2) is 0, A cannot be factorized at ../aten/src/TH/generic/THTensorLapack.cpp:245

In [7]: torch.cholesky_solve(torch.eye(2), a)
Out[7]: 
tensor([[inf, -inf],
        [-inf, inf]])

In [8]: torch.cholesky_solve(torch.eye(2, device='cuda'), a.cuda())
Out[8]: 
tensor([[inf, -inf],
        [-inf, inf]], device='cuda:0')
```

Reply:
Testing (and, in the near future, documenting) this behavior sounds great.